src/compiler/checker.ts

```diff
@@ -2565,6 +2572,14 @@ namespace ts {
                 if (declaration.kind === SyntaxKind.ExportAssignment) {
                     return links.type = checkExpression((<ExportAssignment>declaration).expression);
                 }
+                // Handle module.exports = expr
+                if (declaration.kind === SyntaxKind.BinaryExpression) {
+                    return links.type = checkExpression((<BinaryExpression>declaration).right);
+                }
+                // Handle exports.p = expr
+                if (declaration.kind === SyntaxKind.PropertyAccessExpression && declaration.parent.kind === SyntaxKind.BinaryExpression) {

```

Comment:
Don't think you need the `declaration.parent.kind` test. There should never be a case where that isn't true.